### PR TITLE
Use ConsoleKeyInfo.Key in ConsoleKeyInfo.GetHashCode

### PR DESCRIPTION
**Pull request by stephentoub**
Fixes https://github.com/dotnet/corefx/issues/12380
cc: @weshaggard, @Priya91, @ianhays, @lzybkr

### CODE REVIEW

File: src/System.Console/src/System/ConsoleKeyInfo.cs [hunk 1]
```diff
@@ -70,7 +70,11 @@ public bool Equals(ConsoleKeyInfo obj)
 
         public override int GetHashCode()
         {
-            return (int)_keyChar | (int)_mods;
+            // For all typical cases we can fit all bits losslessly into the hash code:
+            // _keyChar could be any 16-bit value (though is most commonly ASCII). Use 16 bits.
+            // _key can theoretically be 32-bit, but only has enum defined values up to 255.  Use 8 bits.
+            // _mods can also theoretically be 32-bit, but only has enum defined values up to 4. Use 8 bits.
+            return _keyChar | ((int)_key << 16) | ((int)_mods << 24);
```
Comment (lzybkr) [inline src/System.Console/src/System/ConsoleKeyInfo.cs]: Maybe get more bits from `_key` and fewer from `_mods`.  `_key` is the scan code and `254` is actually used, it's `OemClear`.

Comment (stephentoub) [inline src/System.Console/src/System/ConsoleKeyInfo.cs]: > _key is the scan code and 254 is actually used, it's OemClear

I'm not following.  It's effectively taking 8-bits from _key (though still factoring in more if they exist).  254 would fit within that, e.g. 
```C#
char keyChar = (char)0x1234;
ConsoleKey key = ConsoleKey.OemClear;
ConsoleModifiers mods = ConsoleModifiers.Alt | ConsoleModifiers.Control | ConsoleModifiers.Shift;
int hash = keyChar | ((int)key << 16) | ((int)mods << 24);
Console.WriteLine(hash.ToString("X"));
```
prints out:
```
7FE1234
```
What am I not understanding?

Comment (lzybkr) [inline src/System.Console/src/System/ConsoleKeyInfo.cs]: The comment doesn't match the code - the comment says `Use 8 bits`, so fix that for sure. But you're right, you're using 16 bits of `_key` with a potential collision between infrequently used bits of `_key` and frequently used bits of `_mods`.

Given that you are actually using more than 8 bits from `_key`, it probably doesn't matter about potential overlap with `_mods` and using byte aligned shifting might result in better code, so maybe just fix the comment.

Comment (stephentoub) [inline src/System.Console/src/System/ConsoleKeyInfo.cs]: > The comment doesn't match the code - the comment says Use 8 bits, so fix that for sure

I meant 8 bits without any potential for conflict.  I can clarify the comment.

Comment (stephentoub) [inline src/System.Console/src/System/ConsoleKeyInfo.cs]: That said, given this code in the ctor, it'll be extremely uncommon for there to be any value larger than 8-bits:
```C#
        public ConsoleKeyInfo(char keyChar, ConsoleKey key, bool shift, bool alt, bool control)
        {
            // Limit ConsoleKey values to 0 to 255, but don't check whether the
            // key is a valid value in our ConsoleKey enum.  There are a few 
            // values in that enum that we didn't define, and reserved keys 
            // that might start showing up on keyboards in a few years.
            if (((int)key) < 0 || ((int)key) > 255)
            {
                throw new ArgumentOutOfRangeException(nameof(key), SR.ArgumentOutOfRange_ConsoleKey);
            }
```
You'd essentially need to fill in the values of the struct without going through the ctor.

**Comment by stephentoub** [comment]
@dotnet-bot Test Innerloop Ubuntu14.04 Debug Build and Test please (build timeout)

**Pull request merged**